### ISSUE TITLE
Autodetect best syntaxErrorReportingMode based on LSP client?

### ISSUE THREAD

**Issue by michaliskambi** [Isopod/pascal-language-server#3]
While doing https://github.com/Isopod/pascal-language-server/pull/2 , I was testing it also in VS Code, and my conclusion is that every text editor really wants a different value of `syntaxErrorReportingMode` :)

- For VS Code, really the option 1 (window/showMessage) is best. 

    While option 0 (fake completion item, now default) works... it works much worse than window/showMessage: VS Code only shows the fake completion item at some places, and it disables the otherwise sensible fallback of VS Code to do "dummy" dictionary-based completion.

    See the below screens. VS Code shows "fake completion" after `Re` but not after `we` (I did check this was the case also before my PR, so my change, in particular my addition of `isIncomplete` doesn't change VS Code behavior).

    ![Zrzut ekranu z 2022-11-16 04-37-16](https://user-images.githubusercontent.com/1076978/202077922-0660dc12-d36d-4ac7-b2d4-9453b3924790.png)
    ![Zrzut ekranu z 2022-11-16 04-37-28](https://user-images.githubusercontent.com/1076978/202077929-928bf4e3-7925-4cf5-9ac2-9ea74ec798e9.png)

    When errors are reported using `window/showMessage` it is better, message "pops up" regardless if you try to complete after `Re` or `we`. Moreover you get a "dummy" completion, which is better than nothing:
    
    ![Zrzut ekranu z 2022-11-16 04-38-04](https://user-images.githubusercontent.com/1076978/202077992-ffd80d0c-c1ee-46a5-b588-6f4fe5b0e856.png)
    ![Zrzut ekranu z 2022-11-16 04-38-13](https://user-images.githubusercontent.com/1076978/202078012-b02ccdb9-f6a1-466d-a09f-4f627af0b12d.png)

- Note: https://github.com/genericptr/pasls-vscode , the only VS Code integration I know, doesn't for now allow to configure new `syntaxErrorReportingMode`. And I hoped to avoid forking that :)

- For Emacs, both options 0 and 1 are not optimal. So Emacs prefers option 2.

- I don't use NeoVim, but from your comments I understand that NeoVim prefers option 0, as it doesn't display LSP errors and doesn't handle window/showMessage.

All this suggests that syntaxErrorReportingMode should be auto-detected looking at LSP client name, if syntaxErrorReportingMode is not specified in LSP options. We have "clientInfo" given at `initialization` that could be easily used to detect LSP client, VS Code passes `"clientInfo":{"name":"vscode"....}`, Emacs passes `"clientInfo":{"name":"emacs"....}`, I guess NeoVim passes something unique too.

What do you think about this idea? I know that changing behavior based on LSP client name is dirty.. but from user's perspective this is what would make it "just working in the optimal way out-of-the-box".

I could make the autodetection done 

A. when user doesn't specify syntaxErrorReportingMode in LSP options,

B. or when syntaxErrorReportingMode = 0, and make it a default (assumed when not specified in LSP options), thus changing the enum to

```
TSyntaxErrorReportingMode = (
    sermAutoDetect = 0,
    sermFakeCompletionItem = 1,
    sermShowMessage = 2,
    sermErrorResponse = 3
  );
```

B is a bit cleaner IMHO (allows to request "auto-detection" explicitly by "syntaxErrorReportingMode = 0" in LSP options).

(This would break compatibility after https://github.com/Isopod/pascal-language-server/pull/2 , but I guess we have a short window to do this, before anyone except me is really using this option :) ).

If you agree, I could prepare a PR.

**Comment by Isopod** [comment]
I thought about this for a bit and I still haven't decided whether I am for it or against it.

Pro:
- Easy out-of-the box behavior aka "it just works". I like that.

Con:
- Might be a bit too much magic
- We can't possibly cover every client in existence. Seems arbitrary to have special code only for VSCode, Emacs and Neovim.
- Might even have to change behavior based on version number

I think it would be more predictable to choose a compromise that works in most clients. I just retested NeoVim and it seems like `showMessage` is supported now (using nvim 0.8.0 – I originally wrote the code for nvim 0.5.0). It is not ideal, because the message is only visible for ~1 sec, but it's usable (the timeout can probably be tweaked but I didn't look into it). Since the fake completion item causes problems in Emacs and VSCode, we could just change the default to `showMessage`:

- works well in VSCode
- from what I understand, works decently in Emacs
- works decently in revent versions of Neovim
- No magic required

Your thoughts?

**Comment by michaliskambi** [comment]
Agreed, i.e. if we can settle on a default way to show error message, that works decently everywhere, then it's better than auto-detecting LSP clients. Indeed such auto-detection could be "too much magic, too fragile" and we would land in never-ending task of adjusting to all possible versions of all LSP clients.

And yes, I believe this best default should be `showMessage`. It works best in VS Code, it works in NeoVim from what you write, it is passable in Emacs by default. And it seems like a "clean" solution -- we don't report LSP error, we don't pass the message through a fake completion.

Moreover: I decided to "try harder" to get `showMessage` look good in Emacs. The thing is, it is shown perfectly, but immediately after it gets obscured by somewhat useless `No completion found` message from the completion system. And... I found a way to filter out `No completion found` message, and now it is perfect in Emacs! :) The exact EmacsLisp code is on https://github.com/michaliskambi/elisp/blob/master/lsp/kambi-pascal-lsp.el#L60 , based on example "Filtering messages" from https://www.emacswiki.org/emacs/EchoArea#h5o-3 .

![Zrzut ekranu z 2022-11-16 21-47-21](https://user-images.githubusercontent.com/1076978/202292366-8d48036e-24b9-4399-aa56-cc9176e6e47e.png)

So, `showMessage` is actually perfect in Emacs, if one does this extra customization.

So, yes, just changing default to sermShowMessage seems best!
